java-checks/src/main/java/org/sonar/java/checks/LazyArgEvaluationCheck.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.CatchTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Rule(key = "S2629")
+public class LazyArgEvaluationCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final String JAVA_LANG_STRING = "java.lang.String";
+
+  static final class Methods {
+
+    static final class Guava {
+      static final MethodMatcher PRECONDITIONS = MethodMatcher.create()
+        .name("checkState").withAnyParameters();
+
+      private Guava() {}
+    }
+
+    static final class Logging {
+      static final MethodMatcher LOG = MethodMatcher.create()
+        .name("log").addParameter(TypeCriteria.anyType()).addParameter(JAVA_LANG_STRING);
+      static final MethodMatcher TRACE = slf4jPrototype().name("trace");
+      static final MethodMatcher DEBUG = slf4jPrototype().name("debug");
+      static final MethodMatcher INFO = slf4jPrototype().name("info");
+      static final MethodMatcher WARN = slf4jPrototype().name("warn");
+      static final MethodMatcher ERROR = slf4jPrototype().name("error");
+
+      private Logging() {}
+
+      static MethodMatcher slf4jPrototype() {
+        return MethodMatcher.create().addParameter(JAVA_LANG_STRING).addParameter(TypeCriteria.anyType());
```

Comment:
Same remark about owner of the methods here, are you sure that you want to target any method called `trace/debug/info...`. ? 
The FP rate will be pretty high on this I guess. You should  configure the method matcher to cover only methods from `org.slf4j.Logger`.